src/Sylius/Bundle/ThemeBundle/Templating/Locator/TemplateLocator.php

```diff
@@ -0,0 +1,186 @@
+ *
+ * @author Kamil Kokot <kamil.kokot@lakion.com>
+ */
+class TemplateLocator implements FileLocatorInterface
+{
+    /**
+     * @var ThemeRepositoryInterface
+     */
+    private $themeRepository;
+
+    /**
+     * @var ThemeContextInterface
+     */
+    private $themeContext;
+
+    /**
+     * @var ResourceLocatorInterface
+     */
+    private $bundleResourceLocator;
+
+    /**
+     * @var ResourceLocatorInterface
+     */
+    private $applicationResourceLocator;
+
+    /**
+     * @var array
+     */
+    private $cache;
+
+    /**
+     * @param ThemeRepositoryInterface $themeRepository
+     * @param ThemeContextInterface $themeContext
+     * @param ResourceLocatorInterface $bundleResourceLocator
+     * @param ResourceLocatorInterface $applicationResourceLocator
+     * @param string $cacheDir The cache path
+     */
+    public function __construct(
+        ThemeRepositoryInterface $themeRepository,
+        ThemeContextInterface $themeContext,
+        ResourceLocatorInterface $bundleResourceLocator,
+        ResourceLocatorInterface $applicationResourceLocator,
+        $cacheDir = null
+    ) {
+        $this->themeRepository = $themeRepository;
+        $this->themeContext = $themeContext;
+        $this->bundleResourceLocator = $bundleResourceLocator;
+        $this->applicationResourceLocator = $applicationResourceLocator;
+
+        if (null !== $cacheDir && is_file($cache = $cacheDir . '/templates.php')) {
+            $this->cache = require $cache;
+        }
+    }
+
+    /**
+     * Returns a full path for a given file.
+     *
+     * @param TemplateReferenceInterface $template A template
+     * @param string $currentPath Unused
+     * @param bool $first Unused
+     *
+     * @return string The full path for the file
+     *
+     * @throws \InvalidArgumentException When the template is not an instance of TemplateReferenceInterface
+     * @throws \InvalidArgumentException When the template file can not be found
+     */
+    public function locate($template, $currentPath = null, $first = true)
+    {
+        if (!$template instanceof TemplateReferenceInterface) {
+            throw new \InvalidArgumentException("The template must be an instance of TemplateReferenceInterface.");
+        }
+
+        $themes = $this->themeContext->getThemes();
+
+        $templatePath = $this->locateTemplateUsingThemes($template, $themes);
+
+        if (null === $templatePath) {
+            throw new \InvalidArgumentException(
+                sprintf('Unable to find template "%s" (using themes: %s).', $template, join(', ', $themes))
+            );
+        }
+
+        return $templatePath;
+    }
+
+    /**
+     * @param TemplateReferenceInterface $template
+     * @param ThemeInterface[] $themes
+     *
+     * @return null|string
+     */
+    protected function locateTemplateUsingThemes(TemplateReferenceInterface $template, array $themes)
+    {
+        foreach (array_merge($themes, [null]) as $theme) {
```

Comment:
Why is `[null]` here? It seems it's causing bugs.

When doing a cache warmup, first you're looking for the original template path by clearing the context ([line 77 of cache warmer](https://github.com/Sylius/Sylius/pull/2844/files#diff-f865413c726b6bacc90538256bae651bR77)) this causes locator to save a cache entry to the original template. The problem occurs on [line 82 of cache warmer](https://github.com/Sylius/Sylius/pull/2844/files#diff-f865413c726b6bacc90538256bae651bR82)  where you want to locate theme's template path BUT because of `[null]` here it will find the original template from cache therefore instead of **locating** [on line 125 below](https://github.com/Sylius/Sylius/pull/2844/files#diff-70e90ffda6ff082018c7d364aa60322dR125) it will just return original template path from cache.

Two solutions here, forget about the enhancement (of not having duplicate template paths for all themes [by an IF here](https://github.com/Sylius/Sylius/pull/2844/files#diff-f865413c726b6bacc90538256bae651bR84)) and put `$templates[$template->getLogicalName()] = $this->locator->locate($template);` after the for loop, so it doesn't create cache for original tempalte before reading themes templates.

Or just remove `[null]` from here if it is not needed!
Thanks @pamil :D